### PR TITLE
llvm: add support for rebuild and PGO

### PR DESCRIPTION
**Pull request by Andarwinux**
This reduces the first CI build time by 20 minutes. For cached builds, the build system itself and ccache are the biggest bottlenecks. Local users can also easily try LLVM main branch and update LLVM at any time without waiting hours.

### CODE REVIEW

File: toolchain/llvm/llvm.cmake [hunk 1]
```diff
@@ -14,6 +14,7 @@ ExternalProject_Add(llvm
         -DCMAKE_C_COMPILER=clang
         -DCMAKE_CXX_COMPILER=clang++
         ${llvm_ccache}
+        "-DLLVM_THINLTO_CACHE_PATH='${CMAKE_INSTALL_PREFIX}/llvm-thinlto'"
```
Comment (shinchiro) [inline toolchain/llvm/llvm.cmake]: Does CI will save this `llvm-thinlto` dir?

**Comment by shinchiro** [comment]
> This reduces the first CI build time by 20 minutes. For cached builds, the build system itself and ccache are the biggest bottlenecks. 

How about compiling with ccache disabled?

**Comment by zhongfly** [comment]
https://releases.llvm.org/18.1.0/tools/clang/docs/UsersManual.html#profiling-with-instrumentation
According to the documentation, my understanding of the steps to enable PGO for llvm is as follows, please correct me if there are any errors:
1. build llvm with `-fprofile-generate` flag.
2. run llvm with `LLVM_PROFILE_FILE` environment variable, profile data will written to `LLVM_PROFILE_FILE`.`LLVM_PROFILE_FILE` will override `-fprofile-generate=<dir>` setting, so no need to specify profile data dir in  step 1.If we only want to use PGO for llvm, there's no need to build toolchain with `-fprofile-generate`.
3. use profile data to rebuild llvm.I'm not sure if `-fprofile-use` and `-fprofile-generate` can be enabled at the same time, but I'm guessing they can.

**Comment by zhongfly** [comment]
> -fprofile-use and -fprofile-generate are mutually exclusive options.

I haven't found any documentation that says they can't be enabled at the same time, I'll try it later.

**Comment by zhongfly** [comment]
> There does exist another PGO that supports and requires both generate and use, called CSPGO, but that requires three builds, which I haven't tried.

I'm wondering if there's a way to continually optimize iteratively using PGO, currently seems to be possible by continually merging profile data and then building CSPGO?
Build LLVM( -fprofile-generate) → Build LLVM(CSPGO) → LLVM source update → Build LLVM(CSPGO), instead of 'Build LLVM( -fprofile-generate) → Build LLVM( -fprofile-use )' every time.

File: toolchain/llvm/llvm-compiler.in [hunk 2]
```diff
@@ -22,6 +22,9 @@ else
             LTO_FLAGS="$LTO_FLAGS -flto-jobs=@CPU_COUNT@"
         fi
     fi
+    if [ "@LLVM_ENABLE_PGO@" == "GEN" ] || [ "@LLVM_ENABLE_PGO@" == "CSGEN" ]; then
```
Comment (zhongfly) [inline toolchain/llvm/llvm-compiler.in:25]: `if [[ -n "@LLVM_PROFILE_DATA_DIR@" ]]; then`


When training llvm, `LLVM_ENABLE_PGO` is meaningless, no need to set it to `GEN` or `CSGEN`.
 After this modification, training LLVM when building toolchain or packages does not require additional setting of cmake variables.
And just set `LLVM_PROFILE_DATA_DIR` to override the default profile data location.

**Review by zhongfly — CHANGES_REQUESTED** [review]

File: toolchain/llvm/llvm-compiler.in (hunk 2 above)
Comment (zhongfly) [reply toolchain/llvm/llvm-compiler.in:25]: No conflict with `LLVM_PROFDATA_FILE` if `LLVM_ENABLE_PGO=USE`.

**Comment by zhongfly** [comment]
> How about compiling with ccache disabled?

Build mpv,disable ccache, no cache,disable llvm lto,`CLANG_PACKAGES_LTO=ON`:
* Without PGO: 1h
* Training LLVM: 1.5h
* With PGO: 40+min

Haven't compared training with build mpv and just build toolchain yet.

**Comment by zhongfly** [comment]
> Is this CSPGO?

No, CSPGO takes too much time and you replied earlier that the difference is not significant.

Using only the training data from building the toolchain only took 1 minute more compared to merging the data from building all the packages. So it looks like we only need to use the toolchain build to collect data.(`CLANG_PACKAGES_LTO=ON`)

It may take more tries to prove it, but I'm too lazy to do that.

File: ccache_conf.in [hunk 3]
```diff
@@ -1,4 +1,4 @@
 cache_dir = "@MINGW_INSTALL_PREFIX@/ccache"
-max_size = 500M
+max_size = 1G
```
Comment (shinchiro) [inline ccache_conf.in]: Why? 500M big enough for packages building as I dont want the packages' caches grow over time in CI

File: CMakeLists.txt [hunk 4]
```diff
@@ -66,15 +66,32 @@ set(MINGW_INSTALL_PREFIX ${CMAKE_INSTALL_PREFIX}/${TARGET_ARCH} CACHE STRING "Ta
...
+    set(LLVM_PROFDATA_FILE "" CACHE FILEPATH "Profdata file path")
+    if(NOT (EXISTS ${LLVM_PROFDATA_FILE}))
+        set(LLVM_ENABLE_PGO OFF)
+    endif()
+endif()
 option(CLANG_PACKAGES_LTO "Build each package with ThinLTO." OFF)
 option(ENABLE_CCACHE "Enable ccache" OFF)
 if(ENABLE_CCACHE)
+    set(CCACHE_MAXSIZE "500M" CACHE STRING "Size of ccache")
+    set(CCACHE_DIR "${MINGW_INSTALL_PREFIX}/ccache" CACHE STRING "Directory to keep ccached data")
```
Comment (shinchiro) [inline CMakeLists.txt:83]: I dont think `CCACHE_DIR ` option will meaningful as it cannot be shared across multiple target

File: toolchain/llvm/llvm.cmake [hunk 5]
```diff
@@ -142,10 +146,10 @@ ExternalProject_Add(llvm
         -DLLVM_TOOL_VERIFY_USELISTORDER_BUILD=OFF
         -DLLVM_TOOL_VFABI_DEMANGLE_FUZZER_BUILD=OFF
         -DLLVM_TOOL_XCODE_TOOLCHAIN_BUILD=OFF
```
Comment (shinchiro) [inline toolchain/llvm/llvm.cmake:147]: Bring "-DLLVM_THINLTO_CACHE_PATH='${CMAKE_INSTALL_PREFIX}/llvm-thinlto'" before -DCMAKE_C_FLAGS line

File: exec.in [hunk 6]
```diff
@@ -1,10 +1,13 @@
...
+#!/bin/bash
+export O_PATH="$PATH"
 export PATH="@CMAKE_INSTALL_PREFIX@/bin:@RUSTUP_LOCATION@/.cargo/bin:$PATH"
 export PKG_CONFIG="pkgconf --static"
 export PKG_CONFIG_LIBDIR="@MINGW_INSTALL_PREFIX@/lib/pkgconfig"
 export RUSTUP_HOME="@RUSTUP_LOCATION@/.rustup"
 export CARGO_HOME="@RUSTUP_LOCATION@/.cargo"
 export LD_PRELOAD="@MALLOC@"
 export CCACHE_DIR="@MINGW_INSTALL_PREFIX@/ccache"
+export LLVM_PROFILE_FILE="/dev/null"
```
Comment (shinchiro) [inline exec.in:10]: Why this is necessary? `LLVM_PROFILE_FILE` var only exported when  in `GEN` or `CSGEN`. When `OFF` or `USE` it will never export

**Comment by shinchiro** [comment]
Disable thinlto for the CI.

**Comment by shinchiro** [comment]
As usual untested on my local machine but hopefully it does not give me problem later.

**Comment by shinchiro** [comment]
Duh, update requirement in readme too

**Pull request merged**